compiler/src/main/scala/org/scalajs/core/compiler/ExplicitLocalJS.scala

```diff
@@ -0,0 +1,412 @@
+
+import scala.collection.immutable.ListMap
+import scala.collection.mutable
+
+/** Makes explicit the references to local JS classes and desugars calls to
+ *  `js.constructorOf`.
+ *
+ *  It also makes explicit all references to inner JS classes, using the
+ *  pointers created by `ExplicitInnerJS`, and otherwise makes sure the
+ *  back-end will receive all the information it needs to translate inner- and
+ *  local JS classes and objects.
+ *
+ *  Note that in this comment, by "class" we mean *only* `class`es. `trait`s
+ *  and `object`s are not implied.
+ *
+ *  Similarly to how `ExplicitInnerJS` creates explicit fields in the enclosing
+ *  templates of inner JS classes to hold the JS class values, this phase
+ *  creates local vals for local JS classes in the enclosing statement list.
+ *
+ *  For every local JS class of the form:
+ *  {{{
+ *  def outer() = {
+ *    class Local extends ParentJSClass
+ *  }
+ *  }}}
+ *  this phase creates a local `val Local$jslass` in the body of `outer()` to
+ *  hold the JS class value for `Local`. The rhs of that val is a call to a
+ *  magic method, used to retain information that the back-end will need:
+ *
+ *  - A reified reference to `class Local`, in the form of a `classOf`
+ *  - An explicit reference to the super JS class value, i.e., the desugaring
+ *    of `js.constructorOf[ParentJSClass]`
+ *  - An array of fake `new` expressions for all overloaded constructors.
+ *
+ *  The latter will be augmented by `lambdalift` with the appropriate actual
+ *  parameters for the captures of `Local`, which will be needed by the
+ *  back-end. In code, this looks like:
+ *  {{{
+ *  def outer() = {
+ *    class Local extends ParentJSClass
+ *    val Local$jsclass: AnyRef = createLocalJSClass(
+ *        classOf[Local],
+ *        js.constructorOf[ParentJSClass],
+ *        Array[AnyRef](new Local(), ...))
+ *  }
+ *  }}}
+ *
+ *  Since we need to insert fake `new Inner()`s, this scheme does not work for
+ *  abstract local classes. We therefore reject them as implementation
+ *  restriction.
+ *
+ *  If the body of `Local` references itself, then the `val Local$jsclass` is
+ *  instead declared as a `var` to work around the cyclic dependency:
+ *  {{{
+ *  def outer() = {
+ *    var Local$jsclass: AnyRef = null
+ *    class Local extends ParentJSClass {
+ *      ...
+ *    }
+ *    Local$jsclass = createLocalJSClass(...)
+ *  }
+ *  }}}
+ *
+ *  In addition to the above, `ExplicitLocalJS` transforms all *call sites* of
+ *  local JS classes *and* inner JS classes, so that they refer to the
+ *  synthesized local vals and fields.
+ *
+ *  The primary transformation is the desugaring of `js.constructorOf[C]`,
+ *  which depends on the nature of `C`:
+ *
+ *  - If `C` is a statically accessible class, desugar to
+ *    `runtime.constructorOf(classOf[C])` so that the reified symbol survives
+ *    erasure and reaches the back-end.
+ *  - If `C` is an inner JS class, it must be of the form `path.D`, and we
```

Comment:
No here `C := path.D`. If I said `C` is of the form `path.C`, that wouldn't make sense as I would be recursively defining `C := path.C`. So `D`, just like `path`, is whatever fits so that `path.D` gives `C`. In a sense, I'm pattern-matching over `C`, and this is the `case path.C =>` where both `path` and `C` are "captures" from the pattern match.

Maybe it would clearer to add "it must be of the form `path.D` for some pair (`path`, `D`)"?